File: typescript/packages/mechanisms/evm/src/batch-settlement/facilitator/claim.ts
import { SettleResponse, PaymentRequirements } from "@x402/core/types";
import { getAddress } from "viem";
import { FacilitatorEvmSigner } from "../../signer";
import type { AuthorizerSigner, BatchSettlementClaimPayload } from "../types";
import { batchSettlementABI } from "../abi";
import { BATCH_SETTLEMENT_ADDRESS } from "../constants";
import { signClaimBatch } from "../authorizerSigner";
import * as Errors from "../errors";
import { toContractChannelConfig } from "./utils";

/**
 * Converts an array of {@link BatchSettlementVoucherClaim} into the onchain tuple format
 * expected by the contract's `claimWithSignature()` function.
 *
 * @param claims - Typed voucher claims with channel config, amounts, and signatures.
 * @returns Contract-ready VoucherClaim argument array.
 */
export function buildVoucherClaimArgs(claims: BatchSettlementClaimPayload["claims"]) {
  return claims.map(c => ({
    voucher: {
      channel: toContractChannelConfig(c.voucher.channel),
      maxClaimableAmount: BigInt(c.voucher.maxClaimableAmount),
    },
    signature: c.signature,
    totalClaimed: BigInt(c.totalClaimed),
  }));
}

/**
 * Submits a batch claim via `claimWithSignature()`.
 *
 * When `claimAuthorizerSignature` is present in the payload it is used directly.
 * When absent the facilitator signs the `ClaimBatch` EIP-712 digest using
 * `authorizerSigner`, after verifying that every claim's `receiverAuthorizer`
 * matches `authorizerSigner.address`.
 *
 * @param signer - Facilitator signer used to submit the claim transaction.
 * @param payload - Claim payload containing voucher claims and optional authorizer signature.
 * @param requirements - Payment requirements for network identification.
 * @param authorizerSigner - Dedicated key for producing `ClaimBatch` EIP-712 signatures.
 * @returns A {@link SettleResponse} with the transaction hash on success.
 */
export async function executeClaimWithSignature(
  signer: FacilitatorEvmSigner,
  payload: BatchSettlementClaimPayload,
  requirements: PaymentRequirements,
  authorizerSigner: AuthorizerSigner,
): Promise<SettleResponse> {
  const network = requirements.network;
  const claimArgs = buildVoucherClaimArgs(payload.claims);

  let sig = payload.claimAuthorizerSignature;

  if (!sig) {
    for (const claim of payload.claims) {
      if (
        getAddress(claim.voucher.channel.receiverAuthorizer) !==
        getAddress(authorizerSigner.address)
      ) {
        return {
          success: false,
          errorReason: Errors.ErrAuthorizerAddressMismatch,
          transaction: "",
          network,
        };
      }
    }
    sig = await signClaimBatch(authorizerSigner, payload.claims, network);
  }

  try {
    await signer.readContract({
      address: getAddress(BATCH_SETTLEMENT_ADDRESS),
      abi: batchSettlementABI,
      functionName: "claimWithSignature",
      args: [claimArgs, sig],
    });
  } catch (e) {
    return {
      success: false,
      errorReason: Errors.ErrClaimSimulationFailed,
      errorMessage: e instanceof Error ? e.message : String(e),
      transaction: "",
      network,
    };
  }

  try {
    const tx = await signer.writeContract({
      address: getAddress(BATCH_SETTLEMENT_ADDRESS),
      abi: batchSettlementABI,
      functionName: "claimWithSignature",
      args: [claimArgs, sig],
    });

    const receipt = await signer.waitForTransactionReceipt({ hash: tx });

    if (receipt.status !== "success") {
      return {
        success: false,
        errorReason: Errors.ErrClaimTransactionFailed,
        errorMessage: `transaction reverted (receipt status ${receipt.status})`,
        transaction: tx,
        network,
      };
    }

    return {
      success: true,
      transaction: tx,
      network,
      amount: "",
    };
  } catch (e) {
    return {
      success: false,
      errorReason: Errors.ErrClaimTransactionFailed,
      errorMessage: e instanceof Error ? e.message : String(e),
      transaction: "",
      network,
    };
  }
}

File: python/x402/mechanisms/evm/batch_settlement/facilitator/claim.py
"""Facilitator-side claimWithSignature execution."""

from __future__ import annotations

try:
    from eth_utils import to_checksum_address
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .....schemas import PaymentRequirements, SettleResponse
from ...settle_receipt import wait_for_receipt_and_build_response
from ...signer import FacilitatorEvmSigner
from ...utils import truncate_error_message
from ..abi import BATCH_SETTLEMENT_ABI
from ..authorizer_signer import sign_claim_batch
from ..constants import BATCH_SETTLEMENT_ADDRESS
from ..errors import (
    ERR_AUTHORIZER_ADDRESS_MISMATCH,
    ERR_AUTHORIZER_NOT_CONFIGURED,
    ERR_CLAIM_SIMULATION_FAILED,
    ERR_CLAIM_TRANSACTION_FAILED,
)
from ..types import AuthorizerSigner, ClaimPayload, VoucherClaim
from .utils import to_contract_channel_config


def build_voucher_claim_args(claims: list[VoucherClaim]) -> list[tuple]:
    """Convert VoucherClaim list into onchain tuple format for claimWithSignature()."""
    return [
        (
            (
                to_contract_channel_config(c.channel),
                int(c.max_claimable_amount),
            ),
            bytes.fromhex(c.signature.removeprefix("0x")),
            int(c.total_claimed),
        )
        for c in claims
    ]


def execute_claim_with_signature(
    signer: FacilitatorEvmSigner,
    payload: ClaimPayload,
    requirements: PaymentRequirements,
    authorizer_signer: AuthorizerSigner | None,
    data_suffix: str | None = None,
) -> SettleResponse:
    """Submit a batch claim via claimWithSignature()."""
    network = str(requirements.network)
    claim_args = build_voucher_claim_args(payload.claims)

    sig_hex = payload.claim_authorizer_signature
    if not sig_hex:
        if authorizer_signer is None:
            return SettleResponse(
                success=False,
                error_reason=ERR_AUTHORIZER_NOT_CONFIGURED,
                transaction="",
                network=network,
            )
        for claim in payload.claims:
            if to_checksum_address(claim.channel.receiver_authorizer) != to_checksum_address(
                authorizer_signer.address
            ):
                return SettleResponse(
                    success=False,
                    error_reason=ERR_AUTHORIZER_ADDRESS_MISMATCH,
                    transaction="",
                    network=network,
                )
        sig_hex = sign_claim_batch(authorizer_signer, payload.claims, network)

    sig_bytes = bytes.fromhex(sig_hex.removeprefix("0x"))
    target = to_checksum_address(BATCH_SETTLEMENT_ADDRESS)

    try:
        signer.read_contract(
            target, BATCH_SETTLEMENT_ABI, "claimWithSignature", claim_args, sig_bytes
        )
    except Exception as e:
        return SettleResponse(
            success=False,
            error_reason=ERR_CLAIM_SIMULATION_FAILED,
            error_message=truncate_error_message(str(e)),
            transaction="",
            network=network,
        )

    try:
        tx = signer.write_contract(
            target,
            BATCH_SETTLEMENT_ABI,
            "claimWithSignature",
            claim_args,
            sig_bytes,
            data_suffix=data_suffix,
        )
    except Exception as e:
        return SettleResponse(
            success=False,
            error_reason=ERR_CLAIM_TRANSACTION_FAILED,
            error_message=truncate_error_message(str(e)),
            transaction="",
            network=network,
        )

    return wait_for_receipt_and_build_response(
        signer,
        tx,
        network,
        None,
        failed_reason=ERR_CLAIM_TRANSACTION_FAILED,
    )


__all__ = ["build_voucher_claim_args", "execute_claim_with_signature"]
